commons/src/main/java/fr/free/nrw/commons/upload/MwVolleyApi.java

```diff
@@ -0,0 +1,161 @@
+        public void onErrorResponse(VolleyError error) {
+            Log.e(TAG, error.toString());
+        }
+    }
+
+    private static class QueryRequest extends JsonRequest<QueryResponse> {
+        private static final String TAG = QueryRequest.class.getName();
+
+        public QueryRequest(String url,
+                            Response.Listener<QueryResponse> listener,
+                            Response.ErrorListener errorListener) {
+            super(Request.Method.GET, url, null, listener, errorListener);
+        }
+
+        @Override
+        protected Response<QueryResponse> parseNetworkResponse(NetworkResponse response) {
+            String json = parseString(response);
+            //Log.d(TAG, "json=" + json);
+            QueryResponse queryResponse = GSON.fromJson(json, QueryResponse.class);
+            return Response.success(queryResponse, cacheEntry(response));
+        }
+
+
+        private Cache.Entry cacheEntry(NetworkResponse response) {
+            return HttpHeaderParser.parseCacheHeaders(response);
+        }
+
+
+        private String parseString(NetworkResponse response) {
+            try {
+                return new String(response.data, HttpHeaderParser.parseCharset(response.headers));
+            } catch (UnsupportedEncodingException e) {
+                return new String(response.data);
+            }
+        }
+    }
+
+    private static class QueryResponse {
+        private Query query;
+
+        @Override
+        public String toString() {
+            return "query=" + query.toString();
+        }
+    }
+
+    private static class Query {
+        private Page [] pages;
+
+        @Override
+        public String toString() {
+            StringBuilder builder = new StringBuilder("pages=" + "\n");
+            for (Page page : pages) {
+                builder.append(page.toString());
+                builder.append("\n");
+            }
+            builder.replace(builder.length() - 1, builder.length(), "");
+            return builder.toString();
+        }
+    }
+
+    private static class Page {
+        private int pageid;
+        private int ns;
+        private String title;
+        private Category[] categories;
+        private Category category;
+
+        @Override
+        public String toString() {
+
+            StringBuilder builder = new StringBuilder("PAGEID=" + pageid + " ns=" + ns + " title=" + title + "\n" + " CATEGORIES= ");
+            if (categories != null) {
+                for (Category category : categories) {
+                    builder.append(category.toString());
+                    builder.append("\n");
+                }
+            }
+            else {
+                builder.append("no categories exist");
+                builder.append("\n");
+            }
+
+            builder.replace(builder.length() - 1, builder.length(), "");
+            return builder.toString();
+        }
+    }
+
+        private static class Category {
```

Comment:
I was intending to do all of that with private variables within MwVolleyApi... but now that you mention it maybe not such a good idea for me to do that?